modules/workflow-service-impl/src/main/java/org/opencastproject/workflow/impl/WorkflowServiceImpl.java

```diff
@@ -2156,75 +2153,53 @@ public Map<String, Map<String, String>> getWorkflowStateMappings() {
 
   @Override
   public void repopulate(final ElasticsearchIndex index) throws IndexRebuildException {
-    final String startWorkflow = Operation.START_WORKFLOW.toString();
     final int total;
     try {
-      total = serviceRegistry.getJobCount(startWorkflow);
-    } catch (ServiceRegistryException e) {
+      total = persistence.countMediaPackages();
+    } catch (WorkflowDatabaseException e) {
       logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
       throw new IndexRebuildException(index.getIndexName(), getService(), e);
     }
-    final int limit = 1000;
 
     if (total > 0) {
       logIndexRebuildBegin(logger.getSlf4jLogger(), index.getIndexName(), total, "workflows");
+      List<WorkflowIndexData> workflowIndexData;
+
+      try {
+        workflowIndexData = persistence.getWorkflowIndexData();
+      } catch (WorkflowDatabaseException e) {
+        logIndexRebuildError(logger.getSlf4jLogger(), index.getIndexName(), e);
+        throw new IndexRebuildException(index.getIndexName(), getService(), e);
+      }
+
+      Map<String, WorkflowIndexData> latestWorkflows = new HashMap<>();
+      for (WorkflowIndexData data : workflowIndexData) {
+        latestWorkflows.put(data.getMediaPackageId(), data);
+      }
```

Comment:
This works assuming the newest workflow for a given media package comes last.
We could be smarter about it if we reverse the order and then only act whenever we get a new media package id without storing this in a map, overwriting the values again and again.

```
mpid  id
a     5
a     1
b     2
c     4
c     3
```

and the something like:

```py
last_mp = ""
for mp, id in data:
  if mp == last_mp:
    continue
  last_mp = mp
  update_index(data)
```

the SQL query for this would ge something like:

```sql
select id, mediapackage_id, … from oc_workflow order by mediapackage_id, id desc;
```